### PR TITLE
Local run

### PR DESCRIPTION
**Pull request by Lcch**

### CODE REVIEW

File: bucket/bucket.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+// Bucket Object Pool to store objects
+type Bucket interface {
+	OpenRead(key string) (io.Reader, error)
+	OpenWrite(key string) (io.Writer, error)
+}
+
+func IntermidateFileName(mapID int, reduceID int) string {
+	return strconv.Itoa(mapID) + "-" + strconv.Itoa(reduceID) + ".t"
```
Comment (xiang90) [inline bucket/bucket.go]: fmt.Sprintf(xxx) is more readable.

File: executor/executor.go [hunk 2]
```diff
@@ -1,59 +1,198 @@
...
+		inputFile = flag.String("file", "", "input file path")
+		dataDir   = flag.String("intermediate-dir", "/tmp/", "directory of intermediate files")
+		phase     = flag.String("phase", "", "map or reduce")
+		nMap      = flag.Int("nMap", 1, "number of mappers")
+		nReduce   = flag.Int("nReduce", 1, "number of reducers")
+		mapID     = flag.Int("mapID", 0, "mapper id")
+		reduceID  = flag.Int("reduceID", 0, "reducer id")
+	)
+	flag.Parse()
+	if *phase == "map" {
```
Comment (xiang90) [inline executor/executor.go]:
```go

switch *phase {
    case "map":
    case "reduce":
    default:
        panic()
}
```

File: executor/executor.go [hunk 3]
```diff
@@ -1,59 +1,198 @@
...
+		log.Infof("starting id%d mapper", *mapID)
+
+		//ConfigMapper
```
Comment (xiang90) [inline executor/executor.go]: probably remove the comment. it adds no value.

File: executor/executor.go [hunk 4]
```diff
@@ -1,59 +1,198 @@
...
+		//ConfigReducer
+		//
+		bk, err := bucket.NewFilePool(*dataDir + "/" + *jobName)
+		if err != nil {
+			log.Fatalf("Fail to open bucket: %v", err)
+		}
+		// Reduce
+		if res, err := cw.doReduce(bk, *reduceID, *nMap); err != nil {
+			log.Fatalf("Fail to Map: %v", err)
+		} else {
```
Comment (xiang90) [inline executor/executor.go]: remove this else. already fatal when err != nil.

File: executor/executor.go [hunk 5]
```diff
@@ -1,59 +1,198 @@
...
+					inputKV <- r
+				}
+			}
+		}
+	}
+	log.Info("Exit executor")
+}
+
+// Map do Map operation and return aggregated
+func (cw *ComputeWrap) doMap(rr records.RecordReader, bk bucket.Bucket, mapID int, nReduce int) (err error) {
```
Comment (xiang90) [inline executor/executor.go]: doMap -> Map

File: executor/executor.go [hunk 6]
```diff
@@ -1,59 +1,198 @@
...
+		close(waitc)
+	}()
+
+	for rr.HasNext() {
+		inputKV <- ConvertRecordToKV(rr.Pop())
+	}
+	close(inputKV)
+	<-waitc
+
+	log.Debug("DONE Map", time.Now().Sub(startTime))
```
Comment (xiang90) [inline executor/executor.go]: Done Map. Took: , time.Since(start)

File: executor/executor.go [hunk 7]
```diff
@@ -1,59 +1,198 @@
...
+
+	for i := 0; i < nReduce; i++ {
```
Comment (xiang90) [inline executor/executor.go]: this can be ran concurrently?

File: executor/executor.go [hunk 8]
```diff
@@ -1,59 +1,198 @@
...
+		sort.Sort(records.ByKey(aggregated[i]))
+		log.Debug("DONE Map", time.Now().Sub(startTime))
```
Comment (xiang90) [inline executor/executor.go]: this logging line needs more context. (which reduce?)

File: executor/executor.go [hunk 9]
```diff
@@ -1,59 +1,198 @@
...
+		if err != nil {
+			log.Fatalf("Failed to open intermediate: %v", err)
+		}
+		rw := records.MakeRecordWriter("stream", map[string]interface{}{"writer": writer})
+		for _, record := range aggregated[i] {
+			rw.WriteRecord(record)
+		}
+	}
+
+	log.Debug("FINISH Write", time.Now().Sub(startTime))
```
Comment (xiang90) [inline executor/executor.go]: time.Since

File: executor/executor.go [hunk 10]
```diff
@@ -1,59 +1,198 @@
...
+	return
+}
+
+func (cw *ComputeWrap) doReduce(bk bucket.Bucket, reduceID int, nMap int) ([]*kmrpb.KV, error) {
```
Comment (xiang90) [inline executor/executor.go]: doReduce -> reduce

File: records/merge_sorted.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package records
+
+import (
+	"bytes"
```
Comment (xiang90) [inline records/merge_sorted.go]: put an empty line between intern and external imports.

File: util/log/logger.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
+package log
```
Comment (xiang90) [inline util/log/logger.go]: i would suggest to use other's lib instead of creating our own.

File: util/priority_queue.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
+package util
+
+import (
+	"container/heap"
+)
+
+// An Item is something we manage in a priority queue.
+type Item struct {
```
Comment (xiang90) [inline util/priority_queue.go]: probably need a more specific name for this struct. this is not a very generic "Item". We have domain specific fields here, like src and index.

File: util/priority_queue.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+	value interface{}
+	src   int // source of data
+	index int // The index of the item in the heap.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
+type PriorityQueue struct {
+	items   []*Item
+	compare func(a, b interface{}) int
```
Comment (xiang90) [inline util/priority_queue.go]: why a, b are interface{} not Item?

File: executor/executor.go [hunk 15]
```diff
@@ -1,59 +1,198 @@
...
+	readers := make([]records.RecordReader, 0)
+	for i := 0; i < nMap; i++ {
+		reader, err := bk.OpenRead(bucket.IntermidateFileName(i, reduceID))
+		if err != nil {
+			log.Fatalf("Failed to open intermediate: %v", err)
+		}
+		readers = append(readers, records.MakeRecordReader("stream", map[string]interface{}{"reader": reader}))
+	}
+	outputRecords := make([]*kmrpb.KV, 0)
```
Comment (xiang90) [inline executor/executor.go]: this code block looks too dense. need to put some empty line here and there for separation.

File: executor/executor.go [hunk 16]
```diff
@@ -1,59 +1,198 @@
...
+	inputs := make(chan *records.Record, 1024)
+	go records.MergeSort(readers, inputs)
+	var lastKey []byte
+	values := make([][]byte, 0)
+	for r := range inputs {
+		if lastKey == nil || bytes.Compare(lastKey, r.Key) != 0 {
```
Comment (xiang90) [inline executor/executor.go]: i would prefer

```go
if  bytes.Equal(lastKey, r.Key) {
    values = append(values, r.Value)
    continue
}

if lastKey == nil {
    ...
}
...

```

**Comment by comzyh** [comment]
do you forget your rebase yesterday? @Lcch 
https://github.com/comzyh/kmr/commit/11546be43d4b33fe92cced5ae8a84fe9477a2808

File: executor/executor.go [hunk 17]
```diff
@@ -1,59 +1,198 @@
...
+			Value: v,
+		}
+	}
+	close(inputKV)
+	<-waitc
+	return ret, nil
+}
+
+// ConvertRecortToKV
+func ConvertRecordToKV(record *records.Record) *kmrpb.KV {
```
Comment (xiang90) [inline executor/executor.go]: RecordToKV

File: executor/executor.go [hunk 18]
```diff
@@ -1,59 +1,198 @@
...
+	return &kmrpb.KV{Key: record.Key, Value: record.Value}
+}
+
+// ConvertKVtToRecord
+func ConvertKVtToRecord(kv *kmrpb.KV) *records.Record {
```
Comment (xiang90) [inline executor/executor.go]: KVToRecord

File: executor/executor.go [hunk 19]
```diff
@@ -1,59 +1,198 @@
...
+		aggregated = append(aggregated, make([]*records.Record, 0))
+	}
+	waitc := make(chan struct{})
+	inputKV := make(chan *kmrpb.KV, 1024)
+	outputKV := cw.mapFunc(inputKV)
+
+	go func() {
+		for in := range outputKV {
+			rBucketID := util.HashBytesKey(in.Key) % nReduce
+			aggregated[rBucketID] = append(aggregated[rBucketID], ConvertKVtToRecord(in))
```
Comment (xiang90) [inline executor/executor.go]: why do we want to keep ALL output in memory? we should create a pipeline here. keep a batch of data in memory and commit them into disk when we meet the limit (either time or number of output). basically we should combine this goroutine and the routine at line 119.

File: executor/executor.go [hunk 20]
```diff
@@ -1,59 +1,198 @@
...
+	readers := make([]records.RecordReader, 0)
+	for i := 0; i < nMap; i++ {
+		reader, err := bk.OpenRead(bucket.IntermidateFileName(i, reduceID))
+		if err != nil {
+			log.Fatalf("Failed to open intermediate: %v", err)
+		}
+		readers = append(readers, records.MakeRecordReader("stream", map[string]interface{}{"reader": reader}))
+	}
+	outputRecords := make([]*kmrpb.KV, 0)
+	inputs := make(chan *records.Record, 1024)
```
Comment (xiang90) [inline executor/executor.go]: inputs -> sorted

File: executor/executor.go [hunk 21]
```diff
@@ -1,59 +1,198 @@
...
+	go records.MergeSort(readers, inputs)
+	var lastKey []byte
+	values := make([][]byte, 0)
+	for r := range inputs {
```
Comment (xiang90) [inline executor/executor.go]: for r := range sorted

File: executor/executor.go (hunk 10 above)
Comment (xiang90) [inline executor/executor.go]: we need to improve the func in the future. reduce should return a reader of records, not an array of kv. we, now, have to keep all data in memory.

**Comment by xiang90** [comment]
there are still a few minor coding style issues. but i guess we should get this merged as it is for now if it works.

File: master/master.go [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+		}
+	}
+	master.wg.Add(nTasks)
+	master.Unlock()
+	master.wg.Wait()
+
+	if phase == mapPhase {
+		master.commitMappers = make([]int64, 0)
+		for _, t := range master.tasks {
+			master.commitMappers = append(master.commitMappers, t.commitWorker)
```
Comment (comzyh) [inline master/master.go]: I'm a little bit confused.

File: master/master.go [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+
+// RequestTask is to deliver a task to worker.
+func (s *server) RequestTask(ctx context.Context, in *kmrpb.RegisterParams) (*kmrpb.Task, error) {
+	log.Debugf("register %s", in.JobName)
+	s.master.Lock()
+	defer s.master.Unlock()
+
+	for id, t := range s.master.tasks {
+		if t.state == STATE_IDLE {
+			workerID := rand.Int63()
```
Comment (comzyh) [inline master/master.go]: random ID may lead to conflict,  can we use some kind of unique id generator?
Or we add a register method to master, every worker get its unique ID?

File: master/master.go [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+			t.workers[workerID] = id
```
Comment (comzyh) [inline master/master.go]: Why we need this? 
All element in the `workers` field of a specific `Task` struct will have the same value, aka `id`.
Is there some misunderstanding?

File: master/master.go [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+			t.state = STATE_INPROGRESS
+			s.master.heartbeat[workerID] = make(chan int)
+			log.Debug("deliver a task")
+			go s.master.CheckHeartbeatForEachWorker(id, workerID, s.master.heartbeat[workerID])
+			return &kmrpb.Task{
+				WorkerID: workerID,
+				Retcode:  0,
+				Taskinfo: t.taskInfo,
```
Comment (comzyh) [inline master/master.go]: We deliver the `TaskInfo` struct with `commitMappers` to Worker?

File: master/master.go [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+		case kmrpb.ReportInfo_FINISH:
+			heartbeatCode = HEARTBEAT_CODE_FINISH
+		case kmrpb.ReportInfo_DOING:
+			heartbeatCode = HEARTBEAT_CODE_PULSE
+		case kmrpb.ReportInfo_ERROR:
+			heartbeatCode = HEARTBEAT_CODE_DEAD
+		default:
+			panic("unknown ReportInfo")
+		}
+		go func() {
```
Comment (comzyh) [inline master/master.go]: why we need a goroutine here? I think the for loop inside `CheckHeartbeatForEachWorker` function will not consume much time.

File: records/merge_sorted.go [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+	"github.com/naturali/kmr/util"
+)
+
+func MergeSort(readers []RecordReader, outputChan chan<- *Record) error {
+	pq := util.NewPriorityQueue(func(a, b interface{}) int {
+		x, y := a.(*Record), b.(*Record)
+		return bytes.Compare(x.Key, y.Key)
+	})
+	for id, reader := range readers {
```
Comment (comzyh) [inline records/merge_sorted.go]: I used to make a great effort to implement `Peek` method of `ReordReader`. I was thinking we may put `RecordReader` into a heap. now I know we do not need `Peek`.

Thank you.

File: util/priority_queue.go [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+
+import (
+	"container/heap"
+)
+
+// An Item is something we manage in a priority queue.
+type Item struct {
+	value interface{}
+	src   int // source of data
+	index int // The index of the item in the heap.
```
Comment (comzyh) [inline util/priority_queue.go]: Why we need `index`?

according  to https://golang.org/pkg/container/heap/#Interface

If we do not need to update an item in the heap?

File: master/master.go [hunk 29]
```diff
@@ -0,0 +1,238 @@
...
+				master.Unlock()
+				return
+			case HEARTBEAT_CODE_FINISH:
+				master.Lock()
+				if master.tasks[taskID].state != STATE_COMPLETED {
+					master.tasks[taskID].state = STATE_COMPLETED
+					master.tasks[taskID].commitWorker = workerID
+				}
+				master.Unlock()
+				master.wg.Done()
```
Comment (Lcch) [inline master/master.go]: move it inside master.tasks[taskID].state != STATE_COMPLETED

**Pull request merged**